flink-table/flink-table-planner/src/main/java/org/apache/flink/table/plan/rules/logical/PythonCorrelateSplitRule.java

```diff
@@ -0,0 +1,216 @@
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.hep.HepRelVertex;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexLocalRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexProgram;
+import org.apache.calcite.rex.RexProgramBuilder;
+import org.apache.calcite.sql.validate.SqlValidatorUtil;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import scala.Option;
+import scala.collection.Iterator;
+import scala.collection.mutable.ArrayBuffer;
+
+/**
+ * Rule will split the {@link FlinkLogicalTableFunctionScan} which includes java calls into a
+ * {@link FlinkLogicalCalc} which will be the left input of the new  {@link FlinkLogicalCorrelate}
+ * and a new {@link FlinkLogicalTableFunctionScan} without java calls.
+ */
+public class PythonCorrelateSplitRule extends RelOptRule {
+
+	public static final PythonCorrelateSplitRule INSTANCE = new PythonCorrelateSplitRule();
+
+	private PythonCorrelateSplitRule() {
+		super(operand(FlinkLogicalCorrelate.class, any()), "PythonCorrelateSplitRule");
+	}
+
+	private FlinkLogicalTableFunctionScan createNewScan(
+		FlinkLogicalTableFunctionScan scan,
+		ScalarFunctionSplitter splitter) {
+		RexCall rightRexCall = (RexCall) scan.getCall();
+		// extract java funcs.
+		List<RexNode> rightCalcProjects = rightRexCall
+			.getOperands()
+			.stream()
+			.map(x -> x.accept(splitter))
+			.collect(Collectors.toList());
+
+		RexCall newRightRexCall = rightRexCall.clone(rightRexCall.getType(), rightCalcProjects);
+		return new FlinkLogicalTableFunctionScan(
+			scan.getCluster(),
+			scan.getTraitSet(),
+			scan.getInputs(),
+			newRightRexCall,
+			scan.getElementType(),
+			scan.getRowType(),
+			scan.getColumnMappings());
+	}
+
+	@Override
+	public boolean matches(RelOptRuleCall call) {
+		FlinkLogicalCorrelate correlate = call.rel(0);
+		RelNode right = ((HepRelVertex) correlate.getRight()).getCurrentRel();
+		FlinkLogicalTableFunctionScan pythonTableFuncScan;
+		if (right instanceof FlinkLogicalTableFunctionScan) {
+			pythonTableFuncScan = (FlinkLogicalTableFunctionScan) right;
+		} else if (right instanceof FlinkLogicalCalc) {
+			Option<FlinkLogicalTableFunctionScan> scan = CorrelateUtil
+				.getTableFunctionScan((FlinkLogicalCalc) right);
+			if (scan.isEmpty()) {
+				return false;
+			}
+			pythonTableFuncScan = scan.get();
+		} else {
+			return false;
+		}
+		RexNode rexNode = pythonTableFuncScan.getCall();
+		if (rexNode instanceof RexCall) {
+			RexCall pythonTableFuncRexCall = (RexCall) rexNode;
+			return PythonUtil.isPythonCall(pythonTableFuncRexCall) &&
+				pythonTableFuncRexCall.getOperands().stream().anyMatch(PythonUtil::isNonPythonCall);
+		}
+		return false;
+	}
+
+	@Override
+	public void onMatch(RelOptRuleCall call) {
```

Comment:
Could you extract some smaller methods from this method? The code will be more readable.